presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -134,6 +323,139 @@ private static long convertLocalToUtc(DateTimeZone timeZone, long value)
     }
 
+    public static Map<List<String>, ComputedStatistics> createComputedStatisticsToPartitionMap(
+            Collection<ComputedStatistics> computedStatistics,
+            List<String> partitionColumns,
+            Map<String, Type> columnTypes)
+    {
+        List<Type> partitionColumnTypes = partitionColumns.stream()
+                .map(columnTypes::get)
+                .collect(toImmutableList());
+
+        return computedStatistics.stream()
+                .collect(toImmutableMap(statistics -> getPartitionValues(statistics, partitionColumns, partitionColumnTypes), statistics -> statistics));
+    }
+
+    private static List<String> getPartitionValues(ComputedStatistics statistics, List<String> partitionColumns, List<Type> partitionColumnTypes)
+    {
+        checkArgument(statistics.getGroupingColumns().equals(partitionColumns),
+                "Unexpected grouping. Partition columns: %s. Grouping columns: %s", partitionColumns, statistics.getGroupingColumns());
+        Page partitionColumnsPage = new Page(1, statistics.getGroupingValues().toArray(new Block[] {}));
+        return createPartitionValues(partitionColumnTypes, partitionColumnsPage, 0);
+    }
+
+    public static Map<String, HiveColumnStatistics> fromComputedStatistics(
+            ConnectorSession session,
+            DateTimeZone timeZone,
+            Map<ColumnStatisticMetadata, Block> computedStatistics,
+            Map<String, Type> columnTypes,
+            long rowCount)
+    {
+        return createColumnToComputedStatisticsMap(computedStatistics).entrySet().stream()
+                .collect(toImmutableMap(Entry::getKey, entry -> createHiveColumnStatistics(session, timeZone, entry.getValue(), columnTypes.get(entry.getKey()), rowCount)));
+    }
+
+    private static Map<String, Map<ColumnStatisticType, Block>> createColumnToComputedStatisticsMap(Map<ColumnStatisticMetadata, Block> computedStatistics)
+    {
+        Map<String, Map<ColumnStatisticType, Block>> result = new HashMap<>();
+        computedStatistics.forEach((metadata, block) -> {
+            Map<ColumnStatisticType, Block> columnStatistics = result.computeIfAbsent(metadata.getColumnName(), key -> new HashMap<>());
+            columnStatistics.put(metadata.getStatisticType(), block);
+        });
+        return result.entrySet()
+                .stream()
+                .collect(toImmutableMap(Entry::getKey, entry -> ImmutableMap.copyOf(entry.getValue())));
+    }
+
+    private static HiveColumnStatistics createHiveColumnStatistics(
+            ConnectorSession session,
+            DateTimeZone timeZone,
+            Map<ColumnStatisticType, Block> computedStatistics,
+            Type columnType,
+            long rowCount)
+    {
+        HiveColumnStatistics.Builder result = HiveColumnStatistics.builder();
+
+        // MIN MAX
+        if (computedStatistics.containsKey(MIN) && computedStatistics.containsKey(MAX)) {
```

Comment:
does this need to be both or neither?  Sort of makes sense that they'd both be required fields, but it's not totally useless to have only one.  E.g. if min for x was 10 and you had a predicate where x < 8, you'd know that nothing matches.

Reply:
We always compute MIN/MAX in pairs. Checking this in a single if for simplicity. Will add a assertion though.